src/main/java/org/spongepowered/api/item/inventory/InventoryArchetypes.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.inventory;
+
+import org.spongepowered.api.util.generator.dummy.DummyObjectProvider;
+
+/**
+ * An enumeration of {@link InventoryArchetype}s.
+ */
+public class InventoryArchetypes {
+
+    // Base Archetypes
+    // TODO how to start?
+    public static final InventoryArchetype SLOT = DummyObjectProvider.createFor(InventoryArchetype.class, "SLOT");
+    public static final InventoryArchetype MENU_ROW = DummyObjectProvider.createFor(InventoryArchetype.class, "MENU_ROW");
```

Comment:
> ````java
InventoryArchetype menuArchetype = .... ..... .withEvent("button_event_foo") .... .build();
Inventory menuInstance = Inventory.Builder.of(menuArchetype).listener("button_event_foo", Foo::bar).build();
````

well this just looks like my example except that it uses a delegate.
And your `Foo::bar` is a callback (you just name it event. IMO a event uses the `@Listener` approach and `event.getMenuEventId()` or alike.
My example has the benefit that it might be enhenced in the future to support stuff like onDoubleClick, onHover onDragAndDrop, whereas your `withEvent` is strictly limited to one.

> Drop the argument that event handlers are overhead unless you are ready to generate facts to prove it. I'll spare you some effort, they are not.

Yes, you are right they are not overhead, because you have to implement them anyway. (The callback system is build on top of the event system). Its just overhead to force the user to reimplement it every time.